File: sdk/typescript/src/signer.ts
/**
 * HMAC-SHA256 Event Signer for TypeScript
 * Supports both browser (Web Crypto API) and Node.js (crypto module)
 */

export class EventSigner {
  private secret: string;
  private isBrowser: boolean;

  constructor(secret: string) {
    if (!secret) {
      throw new Error('HMAC secret cannot be empty');
    }
    this.secret = secret;
    this.isBrowser = typeof window !== 'undefined' && typeof window.crypto !== 'undefined';
  }

  /**
   * Generate HMAC signature for event.
   */
  async sign(event: any): Promise<string> {
    const canonical = this.canonicalize(event);

    if (this.isBrowser) {
      return this.signBrowser(canonical);
    } else {
      return this.signNode(canonical);
    }
  }

  /**
   * Synchronous signing (Node.js only).
   * In browser, use async sign() instead.
   */
  signSync(event: any): string {
    if (this.isBrowser) {
      throw new Error('Synchronous signing not supported in browser. Use async sign() instead.');
    }
    const canonical = this.canonicalize(event);
    return this.signNode(canonical);
  }

  /**
   * Verify event signature.
   */
  async verify(event: any, signature: string): Promise<boolean> {
    const expected = await this.sign(event);
    return this.constantTimeEqual(expected, signature);
  }

  private canonicalize(event: any): string {
    // Remove signature field and sort keys
    const eventCopy = { ...event };
    delete eventCopy._signature;

    return JSON.stringify(eventCopy, Object.keys(eventCopy).sort());
  }

  private async signBrowser(data: string): Promise<string> {
    const encoder = new TextEncoder();
    const keyData = encoder.encode(this.secret);
    const messageData = encoder.encode(data);

    const key = await crypto.subtle.importKey(
      'raw',
      keyData,
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign']
    );

    const signature = await crypto.subtle.sign('HMAC', key, messageData);

    // Convert ArrayBuffer to hex string
    return Array.from(new Uint8Array(signature))
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
  }

  private signNode(data: string): string {
    // Dynamic import for Node.js crypto
    const crypto = require('crypto');
    const hmac = crypto.createHmac('sha256', this.secret);
    hmac.update(data);
    return hmac.digest('hex');
  }

  private constantTimeEqual(a: string, b: string): boolean {
    if (a.length !== b.length) {
      return false;
    }

    let result = 0;
    for (let i = 0; i < a.length; i++) {
      result |= a.charCodeAt(i) ^ b.charCodeAt(i);
    }

    return result === 0;
  }
}

File: sdk/python/raia/signer.py
"""
Event signer using HMAC-SHA256 for tamper detection.
"""

import hashlib
import hmac
import json
from typing import Any, Dict


class EventSigner:
    """HMAC-SHA256 signer for event integrity."""

    def __init__(self, secret: str):
        """
        Initialize signer with secret key.

        Args:
            secret: HMAC secret key
        """
        if not secret:
            raise ValueError("HMAC secret cannot be empty")
        self.secret = secret.encode("utf-8")

    def sign(self, event: Dict[str, Any]) -> str:
        """
        Generate HMAC signature for event.

        Args:
            event: Event dict to sign

        Returns:
            Hex-encoded HMAC signature
        """
        canonical = self._canonicalize(event)
        signature = hmac.new(self.secret, canonical.encode("utf-8"), hashlib.sha256)
        return signature.hexdigest()

    def verify(self, event: Dict[str, Any], signature: str) -> bool:
        """
        Verify event signature.

        Args:
            event: Event dict
            signature: Hex-encoded signature to verify

        Returns:
            True if signature is valid
        """
        expected = self.sign(event)
        return hmac.compare_digest(expected, signature)

    def _canonicalize(self, event: Dict[str, Any]) -> str:
        """
        Create canonical string representation for signing.
        Excludes _signature field and sorts keys for determinism.
        """
        # Remove signature field if present
        event_copy = {k: v for k, v in event.items() if k != "_signature"}

        # Sort keys recursively for determinism
        return json.dumps(event_copy, sort_keys=True, separators=(",", ":"))


def sign_event(event: Dict[str, Any], secret: str) -> Dict[str, Any]:
    """
    Convenience function to sign an event and add signature field.

    Args:
        event: Event to sign
        secret: HMAC secret

    Returns:
        Event with _signature field added
    """
    signer = EventSigner(secret)
    signature = signer.sign(event)
    event["_signature"] = signature
    return event


def verify_event(event: Dict[str, Any], secret: str) -> bool:
    """
    Convenience function to verify event signature.

    Args:
        event: Event with _signature field
        secret: HMAC secret

    Returns:
        True if signature is valid
    """
    if "_signature" not in event:
        return False

    signature = event["_signature"]
    signer = EventSigner(secret)
    return signer.verify(event, signature)
